### PR TITLE
[CNNDEVOPS-223] `availability_percent` annotaion and Rancher URLs

### PR DESCRIPTION
**Pull request by grahampfau**
This covers two main things:

- Setting an availability threshold via the `availability_percent` annotation. For example, if set to `70` (or `70%`), an alert will not fire if 71% of your pods are available.
- Link to Rancher workload in the alert. `RANCHER_BASE_URL` env var must be defined or `rancher_url` in the alerts will simply be `null`. The controller will still work on non-Rancher managed kubernetes as long as the previously mentioned env var isn't set.

### CODE REVIEW

File: lib/tools/deployments.py [hunk 1]
```diff
@@ -65,6 +83,31 @@ def is_failing(deploy):
...
 
+def meets_threshold(deploy):
+    # meets_threshold, if True an alert should be sent out, if False don't send an alert
+    if not deploy.spec:
+        log.warn('deploy has no spec: {}'.format(deploy.metadata.name))
+        return True
+    annotations = deploy.spec.template.metadata.annotations
+    if annotations and 'availability_percent' in annotations:
+        # availability_percent is in percentages eg. 50 or 99%
+        availability_percent = annotations['availability_percent']
```
Comment (wilsonodk) [inline lib/tools/deployments.py]: I typically would rather have `.strip('%')` in one spot versus having it called every time `availability_percent` is called.

Comment (grahampfau) [reply lib/tools/deployments.py]: Good point. I found a place I can call it only once without creating extra variables. I'll update the PR.

**Comment by abarfoot** [comment]
What event led to this request?  I wonder whether we could instead set a reasonable default.

For example, if the problem is that deployments are causing spurious alerts, we could (if lastUpdateTime is recent or there's another indication that the deployment is in progress), use .spec.strategy.rollingUpdate.maxUnavailable

**Comment by grahampfau** [comment]
@abarfoot As far as I know, this doesn't go off during a rolling upgrade, but it will go off when scaling up the number of replicas in a deployment, but that doesn't happen every day (usually).

**Comment by abarfoot** [comment]
What do you think about -- just never alerting if the lastUpdateTime is less than a minute ago?

I think availability_percent would give people a fine-tuning control that they don't really want.  We'd have to document some guidelines for setting the value, and it seems that the right steady-state value (that is, after the noise of the deployment has settled down) should be 100%...

**Comment by abarfoot** [comment]
(lastUpdateTime might not be the right field to check.  I'm unsure whether it's updated when a deployment changes its "condition".)

**Comment by grahampfau** [comment]
@abarfoot The primary use-case for this is something like Cerebro, where at any given time a pod or two may be unavailable. They have a ticket to investigate and fix that, but until that's done, they can set `availability_percent: 90` until it's fixed to suppress noise, and then the annotation can be removed. Maybe @riyadchowdhury could add in another example.

I'm hesitant to check a field if a deploy just happened, because rolling upgrades don't cause alerts to fire off if the deploy went well, only scale up events. So if there's a data field that indicates a scale up event just happened, ignoring events based off that would be fine, but I don't want ignore real actionable alerts.

**Comment by riyadchowdhury** [comment]
> @abarfoot The primary use-case for this is something like Cerebro, where at any given time a pod or two may be unavailable. They have a ticket to investigate and fix that, but until that's done, they can set `availability_percent: 90` until it's fixed to suppress noise, and then the annotation can be removed. Maybe @riyadchowdhury could add in another example.
> 
> I'm hesitant to check a field if a deploy just happened, because rolling upgrades don't cause alerts to fire off if the deploy went well, only scale up events. So if there's a data field that indicates a scale up event just happened, ignoring events based off that would be fine, but I don't want ignore real actionable alerts.

Another issue is we get tons of flapping alert when a node is being drained. There are some workloads (like Ptolomy) that is running 100 pods, almost one every node. We really shouldn't be alerted on it when one pod dies and gets recreated within minutes. An availability_percent of 90% or something similar would be good for workloads with lots of pids.

**Pull request merged**